lib/protocol/ConnectionManager.js

```diff
@@ -13,52 +13,49 @@
 // language governing permissions and limitations under the License.
 'use strict';
 
-var State = require('./ConnectionManagerState');
 var util = require('../util');
 var debug = util.debuglog('hdb_conn');
 
 module.exports = ConnectionManager;
 
 function ConnectionManager(settings) {
-  this.state = new State(settings);
+  this._settings = settings;
 }
 
-ConnectionManager.prototype.updateState = function updateState(newOptions) {
-  this.state.update(newOptions);
-};
-
-ConnectionManager.prototype.getConnectOptions = function getConnectOptions() {
-  return this.state.options.connect;
+ConnectionManager.prototype.updateConnectOptions = function updateConnectOptions(settings) {
+  return util.extend(this._settings, settings);
 };
 
 ConnectionManager.prototype.openConnection = function openConnection(conn, cb) {
-  var dbHosts = this.state.dbHosts;
+  var s = this._settings;
+  var dbHosts = (Array.isArray(s.hosts) && s.hosts.slice()) || [{'host': s.host, 'port': s.port}]
   debug('Opening connection. Hosts:', dbHosts);
 
-  this._openConnectionToHost(conn, dbHosts, 0, [], cb);
-};
+  var errorStats = [];
+  var self = this;
 
-ConnectionManager.prototype._openConnectionToHost = function _openConnectionToHost(conn, dbHosts, whichHost, errorStats, cb) {
-  if (whichHost === dbHosts.length) {
-    return cb(couldNotOpenConnectionError(errorStats));
+  function _tryConnect() {
+    if (dbHosts.length === 0) {
```

Comment:
:+1: The new way to implement the recursive logic looks more clear than the previous one.